deferral/ah-django
# 165484922 implement server push capability

### What does this PR do?

 - Coverts the server from WSGI To ASGI to facilitate web sockets

### Task to be completed

- [x] Set up an asgi server
- [x] Implement capability for WebSockets

### How can this be manually tested?
To test this feature, you will need a message broker server called Redis. You can install Redis by following  [this tutorial](https://www.digitalocean.com/community/tutorials/how-to-install-and-secure-redis-on-ubuntu-18-04)
- Clone the repo and checkout bg-fix-notifications-165484922
- Install a virtual environment and activate
- Run pip install -r requirements.txt
- Run python manage.py makemigrations
- Run python manage.py migrate
- Run the dev server python manage.py runserver
- Install a WebSocket client of choice or a simple javascript file to will do. [This](https://chrome.google.com/webstore/detail/simple-websocket-client/pfdhoblngboilpfeibdedpjgfnlcodoo?hl=en-US) will do.
- On your selected client enter the following url and hit connect/open ws://127.0.0.1:8000/ws/notifications/<username>/

- Use your favorite HTTP Client to generate notifications by following this [PR](https://github.com/deferral/ah-django/pull/31)

- On your WebSocket client, you will be able to see a JSON object showing the new notification. Similar to the one sent to your email

### Any background context
Unlike email notifications, users have no capability to unsubscribe from in-app notifications.

### PT stories relevant to this PR?
#165484922

### Screenshots
![Screenshot from 2019-04-26 08-53-57](https://user-images.githubusercontent.com/38868860/56786222-e737e700-6800-11e9-859a-33ac1c5f0f6d.png)
